### DISCUSSION TITLE
async vs defer

### DISCUSSION THREAD

**Discussion by oscarbaruffa** [plausible/analytics#1907] [Self-Hosted Support] [answered]
I noticed that the js tracking snippet now has "defer" in place of "async" - should I be refreshing my older sites' js links with this or will it make no difference?

eg <script async data-domain="mydomain.com" src="https://plausible.io/js/plausible.js"></script>

is now
<script defer data-domain="mydomain.com" src="https://plausible.io/js/plausible.js"></script>

**Comment by metmarkosaric** [comment, marked as answer]
you can keep it or change it. there probably won't be much of a difference either way. we had them both in the script at the same time and removed one of them after this post https://github.com/plausible/analytics/discussions/602

**Comment by oscarbaruffa** [comment]
Cool, thanks!

**Comment by cptaffe** [comment]
I used `defer` initially, but ran into an interesting issue. In this case, the Plausible script is held in the _pending_ state for one minute exactly, before failing with `net::ERR_NOT_RESOLVED` [^1]. With the `defer` tag, the request to fetch the script doesn't block parsing of the HTML but it _does_ block the first paint -- so the entire site is left blank if the script fails to load. By contrast, with `async` the site loads as normal regardless of the script failure.

In light of this, I would recommend using `async` as the default -- it leaves the site less vulnerable to issues with the Plausible server.

![Screenshot 2024-05-16 at 8 11 57 PM](https://github.com/plausible/analytics/assets/5462583/536f67a6-29a5-42e3-a97c-f646f48d105c)

[^1]: The root cause here was a network security system, but this behavior could be caused by any DNS-based blocker like Pi-hole as well.

**Reply by matchu** [reply]
+1, I ran into this too, when an outage in my self-hosted Plausible CE instance caused the whole site to stop running Javascript until the analytics script request could time out.

This is because the `defer` attribute not only waits for the document to finish loading, but *preserves the order* of script execution—so scripts that appear later in the `<head>` or `<body>` are presumed to directly depend on this one.

If the script itself isn't depending on the document to be loaded when it runs, and `aysnc` works basically just as well, then I strongly recommend using `async`!
